clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -537,51 +592,151 @@ private void updateGroupSubscription(Set<String> topics) {
+        final Set<TopicPartition> revokedPartitions;
+        final AtomicReference<Exception> firstException = new AtomicReference<>(null);
 
+        // note we should only change the assignment AFTER the callback is triggered
+        // so that users can still access the pre-assigned partitions to commit offsets etc.
         switch (protocol) {
             case EAGER:
-                Set<TopicPartition> revokedPartitions = new HashSet<>(subscriptions.assignedPartitions());
-                log.info("Revoking previously assigned partitions {}", revokedPartitions);
-                try {
-                    listener.onPartitionsRevoked(revokedPartitions);
-                } catch (WakeupException | InterruptException e) {
-                    throw e;
-                } catch (Exception e) {
-                    log.error("User provided listener {} failed on partition revocation", listener.getClass().getName(), e);
-                }
+                // revoke all partitions
+                revokedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+                firstException.compareAndSet(null, maybeInvokePartitionsRevoked(revokedPartitions));
 
-                // also clear the assigned partitions since all have been revoked
                 subscriptions.assignFromSubscribed(Collections.emptySet());
 
                 break;
 
             case COOPERATIVE:
+                // only revoke those partitions that are not in the subscription any more.
+                Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+                revokedPartitions = ownedPartitions.stream()
+                    .filter(tp -> !subscriptions.subscription().contains(tp.topic()))
+                    .collect(Collectors.toSet());
+
+                if (!revokedPartitions.isEmpty()) {
+                    firstException.compareAndSet(null, maybeInvokePartitionsRevoked(revokedPartitions));
+
+                    ownedPartitions.removeAll(revokedPartitions);
+                    subscriptions.assignFromSubscribed(ownedPartitions);
+                }
+
                 break;
         }
 
         isLeader = false;
         subscriptions.resetGroupSubscription();
+
+        if (firstException.get() != null) {
+            throw new KafkaException("User rebalance callback throws an error", firstException.get());
+        }
+    }
+
+    /**
+     * @throws KafkaException if the rebalance callback throws exception
+     */
+    @Override
+    public void resetGeneration(String causeMessage, boolean dueToError) {
+        if (subscriptions.partitionsAutoAssigned()) {
+            // if reset due to error then we have to give up all owned partitions as if they've lost;
+            // otherwise we can give them up as revoked
+            Set<TopicPartition> droppedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+
+            Exception e;
```

Comment:
nit: would be nice to be consistent on the pattern we use here

Reply:
I thought about your suggested API and the other comment from @ableegoldman , and decided to go with the current API which allows for both `atomic reference cas` if multiple callbacks are triggered in order as well as direct assignment when only one callback is needed.